### PR TITLE
HPCC-19494 New IBYTI logic not working properly

### PR DESCRIPTION
**Pull request by richardkchapman**
Code was assuming meaning in subChannels array that was not actually valid.
Make the assumption valid by allocating subChannels in a more cyclic fashion.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Debugged through and observed values and code routes being more what I was expecting.

Note that I don't know if this is the ONLY issue with the new IBYTI code, nor does it explain the memory usage issue which suggests that a Roxie with a high rate of ineffective IBYTI is leaking - this needs to be investigated elsewhere.

### CODE REVIEW

File: roxie/ccd/ccdqueue.cpp [hunk 1]
```diff
@@ -608,51 +607,112 @@ void SlaveContextLogger::flush()
...
+ * this packet is reduced. Any time an IBYTI _does_ arrive on time, the delay is reset to its initial value.
+ */
+
+class SubChannelInfo
 {
-    unsigned mask = SUBCHANNEL_MASK << (SUBCHANNEL_BITS * subChannel);
-    return (liveSubChannels[channel] & mask) == mask;
-}
+public:
+    unsigned getIbytiDelay(unsigned subChannel) const  // NOTE - zero-based
```
Comment (ghalliday) [inline roxie/ccd/ccdqueue.cpp]: Clearer if the parameter was called primarySubChannel, same for the function below.

File: roxie/ccd/ccdqueue.cpp [hunk 2]
```diff
@@ -1489,9 +1549,10 @@ class RoxieReceiverBase : implements IRoxieOutputQueueManager, public CInterface
             assertex(subChannels[channel] == 0);
             assertex(subChannel != 0);
             subChannels[channel] = subChannel;
-            liveSubChannels[channel] = 0;  // Until proven otherwise
             channels[channelCount++] = channel;
         }
+        for (int i = 0; i < channelCount; i++)
+            subChannelInfo[i].init(i);
```
Comment (ghalliday) [inline roxie/ccd/ccdqueue.cpp]: subChannelInfo[channels[i]].init(channels[i]);

or merge into the previous loop

File: roxie/ccd/ccdqueue.cpp [hunk 3]
```diff
@@ -608,51 +607,112 @@ void SlaveContextLogger::flush()
...
-    unsigned mySubChannel = subChannels[h.channel];
+    unsigned primarySubChannel = (hash % numSlaves[h.channel]);
+    unsigned mySubChannel = subChannels[h.channel] - 1;
+    unsigned theirSubChannel = h.getRespondingSubChannel();
+    // could be coded smarter! Basically mysub - prim < theirsub - prim using modulo arithmetic, I think
     while (primarySubChannel != mySubChannel)
     {
-        unsigned channelMask = SUBCHANNEL_MASK << (SUBCHANNEL_BITS * primarySubChannel);
-        if ((h.retries & ROXIE_RETRIES_MASK) == channelMask)
+        if (primarySubChannel == theirSubChannel)
```
Comment (ghalliday) [inline roxie/ccd/ccdqueue.cpp]: still trying to understand this change...

Comment (richardkchapman) [reply roxie/ccd/ccdqueue.cpp]: It was supposed to make this code clearer! Net effect will be the same.

**Pull request merged**